tools/rpmuncompress.c

```diff
@@ -115,28 +116,28 @@ static int singleRoot(const char *path)
 	char *sep = strchr(rootName, '/');
 	if (sep == NULL) {
 	    /* No directories in the pathname */
-	    ret = 0;
+	    rootName = _free(rootName);
 	    goto afree;
 	}
 
+
 	/* Do all entries in the archive start with the same lead directory? */
 	rootLen = sep - rootName + 1;
 	while (archive_read_next_header(a, &entry) == ARCHIVE_OK) {
 	    const char *p = archive_entry_pathname(entry);
 	    if (strncmp(rootName, p, rootLen)) {
-		ret = 0;
+		rootName = _free(rootName);
```

Comment:
Don't free on multiple places, it'll come back to haunt us one way or the other, sooner or later.
AFAICS you could just leave the ret alone and free rootName just before return if ret is anything else than 1?

Reply:
While I generally agree that free-ing at multiple places is bad as it is easy to miss a case and end up with memory leaks, this logic does not really apply here. We return either the string we created once or NULL after freeing. There is no way to create a memory leak here as both options are valid.

Reply:
It's not about leaking now, it's when you inevitably need to change things later. 
Having a single place that frees what needs freeing is the only sane way to do it while in C.

Reply:
Multiple ways to achieve that of course, if you don't like the integer ret, have another pointer that defaults to NULL and only gets set otherwise when single root is detected (ie "assume failure" pattern), that way you can drop the assignments preceding the gotos too.